File: frontend/src/api/casePanel.ts
import type { components } from "./generated/return-platform";
import { APIError, apiClient } from "./client";

/**
 * The case panel, and every review action on it (contracts.md sect. 9).
 *
 * **Types are generated, never mirrored.** Everything below imports from
 * `./generated/return-platform`, which `npm run contracts:generate` emits from
 * the backend's own OpenAPI. A second spelling of `CasePanelView` in this file
 * would be a second thing to keep in step, and the whole point of the contract
 * gate is that there is one.
 *
 * What is *not* generated is the ETag cache and the poll rules further down.
 * Those are decisions about the contract -- when to ask again, what a 304
 * means to a caller who has not kept a copy -- rather than its shape.
 */

/* -------------------------------------------------------------------------
 * The shapes
 * ---------------------------------------------------------------------- */

/**
 * The backend serves every field. `openapi-typescript` renders a Pydantic
 * defaulted field as optional *and* nullable because JSON Schema marks it
 * non-required, but `CasePanelView` is a response model and FastAPI
 * serializes all of it -- so `undefined` is a value this API cannot produce,
 * and carrying it would make every reader narrow a third case that never
 * occurs. Same reasoning as `api/cases.ts`'s `Served<T>`, and the same helper
 * would be imported if it were exported; it is copied deliberately as a local
 * because widening `cases.ts`'s surface for one type alias is the larger
 * change.
 */
type Served<T> = T extends readonly (infer Element)[]
  ? readonly Served<Element>[]
  : T extends object
    ? { [K in keyof T]-?: Served<Exclude<T[K], undefined>> }
    : T;

export type CasePanelView = Served<components["schemas"]["CasePanelView"]>;
export type ReviewPanelView = Served<components["schemas"]["ReviewPanelView"]>;
export type PanelExecutionView = Served<components["schemas"]["PanelExecutionView"]>;
export type PanelTimersView = Served<components["schemas"]["PanelTimersView"]>;
export type PanelSectionView = Served<components["schemas"]["PanelSectionView"]>;
export type AcceptedCommandView = Served<components["schemas"]["AcceptedCommandView"]>;
export type EditStateResult = Served<components["schemas"]["EditStateResult"]>;
export type ReviewActionResult = Served<components["schemas"]["ReviewActionResult"]>;

export type ApproveReviewRequest = components["schemas"]["ApproveReviewRequest"];
export type EditStateRequest = components["schemas"]["EditStateRequest"];
export type ResolveEditRequest = components["schemas"]["ResolveEditRequest"];

/** The states a review can no longer be edited or cancelled from. */
export const IN_FLIGHT_REVIEW_STATES: readonly string[] = [
  "APPROVING",
  "SENT",
  "DELIVERY_FAILED",
  "HELD_FOR_OPERATIONS",
  "ABANDONED",
  "CANCELLED",
];

/** Reviews an associate can still act on. */
export function isEditable(review: ReviewPanelView): boolean {
  return review.state === "OPEN";
}

/** Reviews the recovery actions apply to (contracts.md sect. 6). */
export function isRecoverable(review: ReviewPanelView): boolean {
  return review.state === "DELIVERY_FAILED" || review.state === "HELD_FOR_OPERATIONS";
}

/* -------------------------------------------------------------------------
 * The ETag cache
 * ---------------------------------------------------------------------- */

/**
 * How often an open case's panel is re-read. `copilot.case_poll_interval_ms`
 * in the release; contracts.md sect. 9 makes it the **relay-visibility floor**,
 * which is to say the longest an associate can wait to see that Support
 * answered. Kept equal to the case projection's interval on purpose: two polls
 * for one screen at two cadences would make the panel and the record disagree
 * for up to the difference between them.
 */
export const PANEL_POLL_INTERVAL_MS = 10_000;

/**
 * The last body and its ETag, per case.
 *
 * **Module-scoped rather than per-caller**, because a 304 is only useful to
 * something that kept the bytes: the browser will not replay a body it did not
 * store, and a client that sent `If-None-Match` and then could not answer
 * would have to re-request without it. One case has one panel and every viewer
 * of it in this tab is looking at the same thing, which is exactly the
 * principal-independence the shared payload is built on -- so one entry per
 * case is correct rather than merely convenient.
 *
 * Cleared by `forgetPanel` when a query unmounts, so a long session does not
 * accumulate panels for cases nobody has open.
 */
const cache = new Map<string, { etag: string; view: CasePanelView }>();

export function forgetPanel(caseId: string): void {
  cache.delete(caseId);
}

/** Test seam. Named so its purpose is not mistaken for a cache warmer. */
export function resetPanelCacheForTests(): void {
  cache.clear();
}

type PanelEnvelope = { data?: CasePanelView | null };

/**
 * Read one case's panel, revalidating against the ETag we hold.
 *
 * **`apiClient` cannot be used here and that is the whole reason this
 * function exists.** It reads the JSON body unconditionally, and a 304 has no
 * body -- so the shared client would throw "Unable to read the API response"
 * on the successful half of the mechanism DR-10 chose.
 *
 * A 304 with nothing cached is not an error the caller should see: it means
 * the browser revalidated an entry this module dropped. The request is retried
 * once without the conditional header, which is the only way to get a body
 * back, and the retry is not conditional itself so it cannot loop.
 */
export async function readCasePanel(caseId: string): Promise<CasePanelView> {
  const held = cache.get(caseId);
  const answer = await requestPanel(caseId, held?.etag);
  if (answer.notModified) {
    if (held) return held.view;
    const again = await requestPanel(caseId, undefined);
    if (!again.view) throw new APIError("The case panel could not be read.", again.status);
    remember(caseId, again);
    return again.view;
  }
  if (!answer.view) throw new APIError("The case panel could not be read.", answer.status);
  remember(caseId, answer);
  return answer.view;
}

function remember(
  caseId: string,
  answer: { etag: string | null; view: CasePanelView | null },
): void {
  if (answer.etag && answer.view) cache.set(caseId, { etag: answer.etag, view: answer.view });
}

async function requestPanel(
  caseId: string,
  etag: string | undefined,
): Promise<{
  notModified: boolean;
  status: number;
  etag: string | null;
  view: CasePanelView | null;
}> {
  const headers = new Headers({ Accept: "application/json" });
  if (etag) headers.set("If-None-Match", etag);

  let response: Response;
  try {
    response = await fetch(`/api/v1/cases/${encodeURIComponent(caseId)}/panel`, { headers });
  } catch (error) {
    throw new APIError("Unable to reach the Return Platform API.", 0, undefined, {
      cause: error,
    });
  }

  if (response.status === 304) {
    return { notModified: true, status: 304, etag: response.headers.get("ETag"), view: null };
  }
  if (!response.ok) {
    throw new APIError(
      await panelErrorMessage(response),
      response.status,
      response.headers.get("X-Correlation-ID") ?? undefined,
    );
  }

  const envelope = (await response.json()) as PanelEnvelope;
  return {
    notModified: false,
    status: response.status,
    etag: response.headers.get("ETag"),
    view: envelope.data ?? null,
  };
}

async function panelErrorMessage(response: Response): Promise<string> {
  try {
    const body = (await response.json()) as { detail?: { message?: string } | string };
    if (typeof body.detail === "string") return body.detail;
    if (body.detail?.message) return body.detail.message;
  } catch {
    // A refusal with no JSON body is still a refusal; the status carries it.
  }
  return response.status === 404
    ? "This case is not available."
    : "The case panel could not be read.";
}

/**
 * Whether to keep polling this panel.
 *
 * Stops on a refusal for `caseRefetchInterval`'s reason -- a 403 or 404 will be
 * the same answer in ten seconds, and polling it is a request every ten seconds
 * for as long as the tab is open, to be told the same no.
 *
 * It does **not** stop on "every review is terminal". A case whose reviews have
 * all been sent is still a case the workflow is running, and the panel's
 * execution block, timers and (from V2) parked count keep moving. Stopping
 * there would freeze the screen at the moment the associate handed off.
 */
export function panelRefetchInterval(error?: unknown): number | false {
  if (error instanceof APIError && error.status >= 400 && error.status < 500) return false;
  return PANEL_POLL_INTERVAL_MS;
}

/* -------------------------------------------------------------------------
 * The actions
 * ---------------------------------------------------------------------- */

function reviewPath(caseId: string, reviewId: string, suffix: string): string {
  return `/api/v1/cases/${encodeURIComponent(caseId)}/reviews/${encodeURIComponent(reviewId)}${suffix}`;
}

async function post<T>(path: string, body?: unknown): Promise<T> {
  const response = await apiClient<T>(path, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    ...(body === undefined ? {} : { body: JSON.stringify(body) }),
  });
  if (!response.data) throw new APIError("The action returned no result.", 200);
  return response.data;
}

/**
 * A 409 the UI can act on rather than only report.
 *
 * The backend puts the review's **state** in the body precisely so the panel
 * can say "this review is already approving" instead of "409 Conflict" --
 * contracts.md sect. 6 asks for the transition, and an associate who is shown
 * a status code presses the button again.
 */
export type ReviewConflict = {
  readonly code: string;
  readonly message: string;
  readonly state: string | null;
  readonly field: string | null;
};

export function asReviewConflict(error: unknown): ReviewConflict | null {
  if (!(error instanceof APIError) || error.status !== 409) return null;
  const detail: Record<string, unknown> =
    typeof error.detail === "object" && error.detail !== null
      ? (error.detail as Record<string, unknown>)
      : {};
  return {
    code: typeof detail.code === "string" ? detail.code : "CONFLICT",
    message: error.message,
    state: typeof detail.state === "string" ? detail.state : null,
    field: typeof detail.field === "string" ? detail.field : null,
  };
}

export const casePanelApi = {
  read: readCasePanel,

  /**
   * **This actor's** private edit row. Never in the shared panel body, never
   * in its hash, and served `private, no-store`.
   */
  async readEditState(caseId: string, reviewId: string): Promise<EditStateResult> {
    const response = await apiClient<EditStateResult>(
      reviewPath(caseId, reviewId, "/edit-state"),
    );
    if (!response.data) throw new APIError("The edit state could not be read.", 200);
    return response.data;
  },

  /**
   * One coalesced autosave. `client_edit_id` is the browser's id for this
   * keystroke batch, so a retry over a flaky connection is a no-op rather than
   * a version bump.
   */
  async saveEdit(
    caseId: string,
    reviewId: string,
    body: EditStateRequest,
  ): Promise<EditStateResult> {
    const response = await apiClient<EditStateResult>(
      reviewPath(caseId, reviewId, "/edit-state"),
      {
        method: "PUT",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
      },
    );
    if (!response.data) throw new APIError("The edit could not be saved.", 200);
    return response.data;
  },

  /** Select, merge or discard, resolved to one canonical payload. */
  resolveEdit(
    caseId: string,
    reviewId: string,
    body: ResolveEditRequest,
  ): Promise<ReviewActionResult> {
    return post<ReviewActionResult>(reviewPath(caseId, reviewId, "/edit-state/resolve"), body);
  },

  /** The three CAS values are the ones the associate read, never the store's. */
  approve(
    caseId: string,
    reviewId: string,
    body: ApproveReviewRequest,
  ): Promise<ReviewActionResult> {
    return post<ReviewActionResult>(reviewPath(caseId, reviewId, "/approve"), body);
  },

  revise(caseId: string, reviewId: string, note?: string): Promise<ReviewActionResult> {
    return post<ReviewActionResult>(reviewPath(caseId, reviewId, "/revise"), {
      note: note ?? null,
    });
  },

  cancel(caseId: string, reviewId: string, reason: string): Promise<ReviewActionResult> {
    return post<ReviewActionResult>(reviewPath(caseId, reviewId, "/cancel"), { reason });
  },

  redraft(caseId: string, reviewId: string): Promise<ReviewActionResult> {
    return post<ReviewActionResult>(reviewPath(caseId, reviewId, "/template-review/redraft"));
  },

  retryDelivery(
    caseId: string,
    reviewId: string,
    reason = "",
  ): Promise<ReviewActionResult> {
    return post<ReviewActionResult>(reviewPath(caseId, reviewId, "/recovery/retry"), { reason });
  },

  abandon(caseId: string, reviewId: string, reason: string): Promise<ReviewActionResult> {
    return post<ReviewActionResult>(reviewPath(caseId, reviewId, "/recovery/abandon"), {
      reason,
    });
  },
};

/** Query keys, in one place so an invalidation cannot miss a reader. */
export const casePanelKeys = {
  panel: (caseId: string) => ["case-panel", caseId] as const,
  editState: (caseId: string, reviewId: string) =>
    ["case-panel", caseId, "edit-state", reviewId] as const,
};

File: backend/src/return_platform/operations/case_panel.py
"""`CasePanelView` and the section registry (contracts.md sect. 9, DR-10).

**This module's contract is frozen at merge.** V1 ships the DTO and the
registry and then never touches either again; V2 and V3 contribute their
sections through `register_panel_section` in their own files. That is the whole
point of the seam -- three slices adding fields to one DTO is three slices
editing one file, and the last one to merge wins an argument nobody had.

---

**What the shared payload is, and what it deliberately is not.**

Every field here is *case-scoped and principal-independent*. Two principals
who may both read a case get byte-identical bodies and therefore identical
ETags, which is what makes a shared cache honest and is asserted directly.
`accepted_commands[]` is unfiltered for the same reason: filtering it by actor
would make the hash actor-dependent while looking like a privacy improvement,
and the actor ids in it are already visible in the review's state history.

Per-actor edit state is **not here**. It lives at
`GET .../reviews/{review_id}/edit-state`, `Cache-Control: private, no-store`,
because an autosaved draft is one person's unfinished thinking and putting it
in a shared, cacheable, hashed body would publish it to everyone who can read
the case. The *conflict marker* participates in the hash -- "somebody else is
editing this" is a case-level fact -- while the edit contents never do.

**Absolute instants only.** `deadline_iso` is an instant; there is no
`seconds_remaining`. A server-computed countdown is stale the moment it is
serialized, and worse, it makes the hash change every second so no ETag ever
matches. The browser counts down.

**Degradation is narrow.** A section answers `{status: "degraded", reason}`
only for an expected timeout or a transient failure -- a Temporal query that
timed out, a projection that could not be assembled. An authorization failure,
a contract violation or a broken invariant is a real error and is raised as
one: a panel that renders "degraded" over a 403 is a panel that hides a
security answer behind a spinner. And a degraded section is **never a
`TemplateGap`** -- gaps come only from the renderer, and mean the *case* does
not know something, not that we could not read it.
"""

from __future__ import annotations

from collections.abc import Callable, Mapping, Sequence
from typing import Any, Final, Protocol

from pydantic import BaseModel, ConfigDict, Field

__all__ = [
    "PANEL_SECTION_IDS",
    "AcceptedCommandView",
    "CasePanelView",
    "PanelExecutionView",
    "PanelSectionContributor",
    "PanelSectionView",
    "PanelTimersView",
    "ReviewPanelView",
    "canonical_panel_payload",
    "clear_panel_sections",
    "panel_etag",
    "panel_section_contributors",
    "register_panel_section",
    "sorted_sections",
]


class _Panel(BaseModel):
    """Order-stable and closed. Both matter to the hash.

    `extra="forbid"` so a section cannot smuggle a field past the contract, and
    field order is declaration order, which is what makes the canonical
    serialization below deterministic without a sorting pass over every nested
    object.
    """

    model_config = ConfigDict(extra="forbid", frozen=True)


class PanelSectionView(_Panel):
    """One contributed section. **V2 and V3 add sections, never fields.**

    The payload is an opaque JSON object rather than a typed sub-model, and
    that is the seam: a typed field per section would put V2's and V3's shapes
    into this file and into every OpenAPI regeneration V1 owns. Each
    contributor owns its own payload's shape and its own tests for it.
    """

    section_id: str
    #: `"ok"` or `"degraded"`. A degraded section still renders -- with its
    #: reason -- rather than vanishing, because a section that disappears looks
    #: to an associate exactly like a case that has nothing to show.
    status: str = "ok"
    reason: str | None = None
    payload: dict[str, Any] = Field(default_factory=dict)


class PanelExecutionView(_Panel):
    """The Temporal `execution_state` query, as the panel needs it.

    Degradable: the workflow host being unreachable is an expected transient,
    and the rest of the panel -- reviews, records, the thread digest -- is read
    from Mongo and is still true.
    """

    status: str = "ok"
    reason: str | None = None
    case_status: str | None = None
    work_item_id: str | None = None
    awaiting: tuple[str, ...] = ()
    business_complete: bool = False
    parked_reason: str | None = None


class ReviewPanelView(_Panel):
    """One review, keyed `(review_kind, scope_id)` (contracts.md sect. 9).

    Carries every non-terminal review *including* `APPROVING`,
    `DELIVERY_FAILED` and `HELD_FOR_OPERATIONS` -- a review the associate can
    no longer edit is precisely the one they most need to see -- plus recently
    terminal ones for visibility.

    `draft` is the payload as it currently stands; `gaps` come only from the
    renderer. `conflict_present` is the case-scoped marker's answer for this
    review, never a per-actor fact.
    """

    review_id: str
    review_kind: str
    scope_id: str
    request_id: str
    state: str
    draft_version: int
    canonical_edit_version: int
    conflict_present: bool
    draft: dict[str, Any] = Field(default_factory=dict)
    gaps: tuple[dict[str, Any], ...] = ()
    #: The digest of the payload approval would freeze, as the store holds it
    #: **right now**.
    #:
    #: **Served rather than left for the client to derive, and that is the whole
    #: point of it.** Approval's compare-and-set requires
    #: `canonical_approved_payload_hash` to equal
    #: `canonical_payload_digest(canonical_review_payload(review))` -- over the
    #: store's canonical serialization, of the canonical edit where one exists
    #: and the draft where it does not. A browser computing that would be a
    #: second implementation of the CAS, in another language, and the two would
    #: disagree the first time either side changed how a payload serializes.
    #: Every approval from the console would then answer 409 for a reason no
    #: associate could act on.
    #:
    #: Echoing it back loses nothing the CAS is for: the guarantee is that an
    #: associate approves *the bytes they read*, and a draft that moved between
    #: this panel read and their approval produces a different digest, so the
    #: store still refuses. `None` only when the review is past `OPEN` and there
    #: is nothing left to approve.
    approval_hash: str | None = None
    #: `None` unless the review has been in `APPROVING` or past it. The panel
    #: shows "approved by X, sending" from this, and it is server-stamped.
    approved_by: str | None = None
    approved_at_iso: str | None = None
    #: Present from `DELIVERY_FAILED` on. What the recovery actions act on.
    recovery_status: str | None = None
    last_delivery_error_code: str | None = None
    hold_reason: str | None = None
    abandon_audit: dict[str, Any] | None = None


class AcceptedCommandView(_Panel):
    """A command the platform accepted and has not necessarily applied yet.

    **Unfiltered by actor** -- see the module docstring. It is what makes "I
    pressed Send and nothing happened" answerable: the command is recorded, the
    signal has not landed yet, and the panel can say so instead of showing an
    unchanged review.
    """

    signal_id: str
    kind: str
    actor_id: str
    review_id: str | None = None
    recorded_at_iso: str | None = None
    applied: bool = False


class PanelTimersView(_Panel):
    """Absolute instants. The countdown is the browser's (contracts.md sect. 9)."""

    template_review_deadline_iso: str | None = None
    template_review_reminders_sent: int = 0
    template_review_max_reminders: int = 0
    support_deadline_iso: str | None = None


class CasePanelView(_Panel):
    """The one shape the panel and `CaseOperationsPage` both read.

    **Frozen at merge.** Anything V2 or V3 wants to show goes in `sections`
    through the registry.

    That sentence used to be false for three fields. `support_digest`,
    `clarifications` and `parked_messages` were declared here as top-level
    placeholders for contributing slices to fill -- but a contributor satisfies
    `PanelSectionContributor`, which returns a `PanelSectionView | None` into
    `sections`, and **has no way to write a top-level field**. So the composer
    hardcoded all three empty and no contributor could ever change that. V3
    built a clarifications section against `panel.clarifications`; it would have
    drawn nothing on every real panel while a suite full of hand-built panel
    objects stayed green. AMENDMENT-6 retired them, and this is where that
    lands.

    **Do not re-add a top-level field for a contributing slice.** If a section
    needs to say something, it says it in its own `PanelSectionView.payload`.
    A second parallel path that the seam cannot reach is the defect, not the
    absence of one.
    """

    case_id: str
    execution: PanelExecutionView
    reviews: tuple[ReviewPanelView, ...] = ()
    return_records: tuple[dict[str, Any], ...] = ()
    timers: PanelTimersView = Field(default_factory=PanelTimersView)
    accepted_commands: tuple[AcceptedCommandView, ...] = ()
    #: Contributed sections, **sorted by `section_id`**. Sorted rather than
    #: registration-ordered because registration order depends on import order,
    #: and an ETag that changed when a module moved would be a cache that
    #: misses for a reason nobody can see.
    sections: tuple[PanelSectionView, ...] = ()


# --------------------------------------------------------------------------- #
# The registry
# --------------------------------------------------------------------------- #


class PanelSectionContributor(Protocol):
    """What V2 and V3 implement.

    `context` is deliberately a plain mapping rather than a typed object: the
    contributors live in other slices, and a typed context would be a second
    frozen contract this file owns. What it holds is documented at
    `register_panel_section`.
    """

    async def __call__(self, context: Mapping[str, Any]) -> PanelSectionView | None: ...


_CONTRIBUTORS: dict[str, PanelSectionContributor] = {}

#: Section ids V1 knows about, for the reader. Contributors are **not**
#: restricted to this list -- a registry that had to be edited to accept a new
#: section would be the frozen DTO again, wearing a dict.
PANEL_SECTION_IDS: Final[tuple[str, ...]] = ()


def register_panel_section(section_id: str, contributor: PanelSectionContributor) -> None:
    """Contribute one section to every case panel.

    Called at import time from the owning slice's own module. Registering the
    same id twice is a programming error and raises: two contributors for one
    section would race for the id, and which one won would depend on import
    order.

    The `context` a contributor receives carries at minimum `case_id`,
    `tenant_id`, `principal_id` and `request` (the FastAPI `Request`, so a
    contributor can reach its own app state). A contributor that needs
    something else asks for it here rather than reaching into the panel
    composer, which is what keeps this seam one-directional.

    Returning `None` omits the section. Raising is caught by the composer and
    becomes a **degraded** section -- a contributor must not be able to take
    the whole panel down, because the reviews are the part an associate is
    actually blocked on.
    """
    if section_id in _CONTRIBUTORS:
        raise ValueError(f"panel section {section_id!r} is already registered")
    _CONTRIBUTORS[section_id] = contributor


def panel_section_contributors() -> tuple[tuple[str, PanelSectionContributor], ...]:
    """Every registered contributor, **sorted by id**. See `CasePanelView.sections`."""
    return tuple(sorted(_CONTRIBUTORS.items()))


def clear_panel_sections() -> None:
    """Empty the registry. For tests, and named so that is obvious."""
    _CONTRIBUTORS.clear()


def _canonical(value: Any) -> Any:
    """Order-stable JSON, so one body always hashes to one ETag.

    Mappings are emitted in sorted key order and every other container keeps
    its own order, which is meaningful -- `reviews[]` is oldest-first and
    reordering it would be a different panel. Bytes are refused rather than
    coerced: nothing in this DTO holds them, and a silent `str()` of one would
    hash two different values the same.
    """
    if isinstance(value, Mapping):
        return {str(key): _canonical(value[key]) for key in sorted(value, key=str)}
    if isinstance(value, (list, tuple)):
        return [_canonical(item) for item in value]
    if isinstance(value, (bytes, bytearray)):
        raise TypeError(
            "the panel payload holds no bytes, and coercing them would hash two "
            "different values the same"
        )
    return value


def canonical_panel_payload(view: CasePanelView) -> dict[str, Any]:
    """The bytes the ETag is computed over."""
    return _canonical(view.model_dump(mode="json"))


def panel_etag(view: CasePanelView, *, digest: Callable[[Mapping[str, Any]], str]) -> str:
    """`ETag` for one composed panel (DR-10).

    The digest function is injected rather than imported so this module stays
    free of the support-event machinery; the composer passes
    `canonical_payload_digest`, which is the same digest the approval hash uses.
    Quoted, because an unquoted ETag is not a valid one and some caches drop it.
    """
    return f'"{digest(canonical_panel_payload(view))}"'


def sorted_sections(sections: Sequence[PanelSectionView]) -> tuple[PanelSectionView, ...]:
    return tuple(sorted(sections, key=lambda section: section.section_id))
